Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/D7AbsoluteCrossSections.py

```diff
@@ -0,0 +1,504 @@
+                            sigma_z_sf = 0.5 * (sigma_z_sf + mtd[second_xyz_ws][entry_no].readY(spectrum))
+                            sigma_z_nsf = 0.5 * (sigma_z_nsf + mtd[second_xyz_ws][entry_no+1].readY(spectrum))
+                            sigma_xmy_sf = mtd[second_xyz_ws][entry_no+2].readY(spectrum)
+                            sigma_xmy_nsf = mtd[second_xyz_ws][entry_no+3].readY(spectrum)
+                            sigma_xpy_sf = mtd[second_xyz_ws][entry_no+4].readY(spectrum)
+                            sigma_xpy_nsf = mtd[second_xyz_ws][entry_no+5].readY(spectrum)
+                        # Magnetic component
+                        theta_0 = DEG_2_RAD * self._sampleAndEnvironmentProperties['ThetaOffset'].value
+                        theta = mtd[ws][entry_no].detectorInfo().twoTheta(spectrum)
+                        alpha = theta - 0.5*np.pi - theta_0
+                        c0 = math.pow(math.cos(alpha), 2)
+                        c4 = math.pow(math.cos(alpha - np.pi/4.0), 2)
+                        magnetic_nsf_cos2alpha = (2*c0-4)*sigma_x_nsf + (2*c0+2)*sigma_y_nsf + (2-4*c0)*sigma_z_nsf
+                        magnetic_sf_cos2alpha = -(2*c0-4)*sigma_x_sf - (2*c0+2)*sigma_y_sf - (2-4*c0)*sigma_z_sf
+                        magnetic_nsf_sin2alpha = (2*c4-4)*sigma_xpy_nsf + (2*c4+2)*sigma_xmy_nsf + (2-4*c4)*sigma_z_nsf
+                        magnetic_sf_sin2alpha = -(2*c4-4)*sigma_xpy_sf - (2*c4+2)*sigma_xmy_sf - (2-4*c4)*sigma_z_sf
+                        dataY_nsf_magnetic[spectrum] = magnetic_nsf_cos2alpha * math.cos(2*alpha) \
+                            + magnetic_nsf_sin2alpha * math.sin(2*alpha)
+                        dataY_sf_magnetic[spectrum] = magnetic_sf_cos2alpha * math.cos(2*alpha) + magnetic_sf_sin2alpha * math.sin(2*alpha)
+                        dataY_average_magnetic[spectrum] = 0.5 * (dataY_nsf_magnetic[spectrum] + dataY_sf_magnetic[spectrum])
+                        # Nuclear coherent component
+                        dataY_nuclear[spectrum] = (2.0 * (sigma_x_nsf + sigma_y_nsf + 2*sigma_z_nsf + sigma_xpy_nsf + sigma_xmy_nsf)
+                                                   - (sigma_x_sf + sigma_y_sf + 2*sigma_z_sf + sigma_xpy_sf + sigma_xmy_sf)) / 12.0
+                        # Incoherent component
+                        dataY_incoherent[spectrum] = 0.25 * (sigma_x_sf + sigma_y_sf + 2*sigma_z_sf + sigma_xpy_sf + sigma_xmy_sf) \
+                            - dataY_average_magnetic[spectrum]
+
+            dataY = [dataY_total, dataY_nuclear, dataY_incoherent, dataY_average_magnetic, dataY_nsf_magnetic, dataY_sf_magnetic]
+            if user_method == 'XYZ' or user_method == '10p':
+                nCrossSections = 6
+            else:
+                nCrossSections = nComponents
+            for cross_section_no in range(nCrossSections):
+                dataX = mtd[ws][entry_no].readX(0)
+                dataE = np.sqrt(abs(dataY[cross_section_no]))
```

Comment:
I see you are extracting the spectra and making arithmetics with those and setting the uncertainty as the square root of the resulting intensity. This is wrong for reduced data. You should stick to workspaces and perform all the operations as workspaces, that way the uncertainties will be correctly propagated through.